### PR TITLE
(Fix #254) Support alternate locations for LLVM on Windows

### LINKED ISSUE #254: Unable to download prebuilt binaries or complete the full build of skia.
When trying to build Skulpin I'm getting the following error (the tarball for the 0.23 release is being incorrectly determined)
```
HOST: x86_64-pc-windows-msvc
TRYING TO DOWNLOAD AND INSTALL SKIA BINARIES: 0.23.0/bf61a03a26d021e125fb-x86_64-pc-windows-msvc-svg-textlayout-vulkan-static
  FROM: https://github.com/rust-skia/skia-binaries/releases/download/0.23.0/skia-binaries-bf61a03a26d021e125fb-x86_64-pc-windows-msvc-svg-textlayout-vulkan-static.tar.gz
DOWNLOAD AND INSTALL FAILED: [22] HTTP response code said error
```

When trying to build rust-skia from the repo I get a different error about a missing git ref:
```
--- stderr
Submodule 'skia-bindings/depot_tools' (https://github.com/rust-skia/depot_tools.git) registered for path 'depot_tools'
Submodule 'skia' (https://github.com/google/skia.git) registered for path 'skia'
Cloning into 'C:/projects/rust-skia/skia-bindings/depot_tools'...
Cloning into 'C:/projects/rust-skia/skia-bindings/skia'...
From https://github.com/google/skia
 * branch            2542bdfcd686c7bfa9f687d80627f8ef468b2838 -> FETCH_HEAD
warning: remote HEAD refers to nonexistent ref, unable to checkout.

ninja: fatal: ReadFile: The handle is invalid.
```

I don't know exactly how to work this out as this is the first interaction with the system I've had so far. Any tips?

### PR DESCRIPTION
**Pull request by photex**
Here is at least a first possible solution for my situation. I'm happy to make any changes required.

I tested this with an example from Skulpin.

### CODE REVIEW

File: skia-bindings/build_support/skia.rs [hunk 1]
```diff
@@ -295,8 +302,34 @@ impl FinalBuildConfiguration {
...
-                    args.push(("clang_win", quote("C:/Program Files/LLVM")));
+                    // Attempt to locate the LLVM installation using a list of common prefixes
+                    // otherwise use the environment variable LLVM_HOME.
+                    let llvm_home = if let Ok(llvm_home) = env::var("LLVM_HOME") {
+                        if !check_llvm_root(&llvm_home) {
+                            panic!("Unable to locate clang-cl.exe in {}. Unable to build skia-bindings", llvm_home);
+                        }
+                        llvm_home
+                    } else {
+                        let userprofile = env::var("USERPROFILE").expect("Unable to resolve %USERPROFILE%");
```
[the remark this answers is not in the document]
Comment (pragmatrix) [reply skia-bindings/build_support/skia.rs]: Ah, it's `cargo::env_var()`, not `cargo::var()` sorry.

**Comment by photex** [comment]
Super good points. I'll get them sorted out in a jiffy. :D

File: skia-bindings/build_support/skia.rs [hunk 2]
```diff
@@ -295,8 +302,34 @@ impl FinalBuildConfiguration {
...
+                        let common_roots = [
+                            String::from("C:\\Program Files\\LLVM"),
+                            String::from("C:\\LLVM"),
+                            format!("{}\\scoop\\apps\\llvm\\current", userprofile),
+                        ];
+                        let mut llvm_home = String::new();
```
[the remark this answers is not in the document]
Comment (photex) [reply skia-bindings/build_support/skia.rs]: Yerp! I never do this, but was in a hurry to get this PR rolling before picking my kid up from a birthday and didn't really pay close enough attention haha.

File: skia-bindings/build_support/skia.rs [hunk 3]
```diff
@@ -295,8 +302,34 @@ impl FinalBuildConfiguration {
...
+                                eprintln!("Found clang-cl.exe in {}", root);
+                                llvm_home = root.clone();
+                                break;
+                            }
+                        }
+                        if llvm_home.is_empty() {
+                            panic!("Unable to locate clang-cl.exe.");
+                        }
+                        llvm_home
+                    };
```
[the remark this answers is not in the document]
Comment (photex) [reply skia-bindings/build_support/skia.rs]: Seems to make sense to put it in llvm.rs given that it's specifically looking for an llvm installation on windows.
Do you prefer to have a platform submodule? `llvm::win` or `llvm::windows`?

File: skia-bindings/build_support/skia.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (photex) [reply skia-bindings/build_support/skia.rs]: That's awesome. I had no idea cargo could even do this.... I have so many build scripts to update now ha!

File: skia-bindings/build_support/skia.rs [hunk 4]
```diff
@@ -223,6 +223,13 @@ impl FinalBuildConfiguration {
                 format!("\"{}\"", s)
             }
 
+            fn check_llvm_root(root: &str) -> bool {
+                let mut clang_cl = std::path::PathBuf::from(root);
+                clang_cl.push("bin\\clang-cl.exe");
+                eprintln!("Checking {}", clang_cl.to_str().unwrap());
```
[the remark this answers is not in the document]
Comment (photex) [reply skia-bindings/build_support/skia.rs]: I honestly don't know, but I just assumed this was the case. I only expect to see the output when there is an issue normally, so in addition to the panic message I look at the stderr output for help with fixing problems.
Should I just remove these lines?

File: skia-bindings/build_support/skia.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (photex) [reply skia-bindings/build_support/skia.rs]: I went with `llvm::win` for now. I can move it up if you like.

File: skia-bindings/build_support/skia.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pragmatrix) [reply skia-bindings/build_support/skia.rs]: No leave them as they are, both, stderr and stdout are only shown when `-vv` is used in cargo invocations like in `cargo build -vv`. I'll decide later if they need to be consistent or if they should be classified by severity.

File: skia-bindings/build_support/skia.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pragmatrix) [reply skia-bindings/build_support/skia.rs]: `llvm::win` is perfect!

**Comment by pragmatrix** [comment]
BTW: Nicely done, and a great pleasure working with you. I hope you can use rust-skia for what you are planning next.

**Comment by photex** [comment]
Thanks! And likewise.

I have high hopes!

**Pull request merged**